### PR TITLE
Fix camlinternalOO at -O3 with Flambda 2

### PR DESCRIPTION
**Pull request by mshinwell**
`CamlinternalOO` contains various unsafe operations, some of which seem to confuse arrays and blocks, to which Flambda 2 takes exception.  This PR attempts to improve matters.

### CODE REVIEW

File: ocaml/stdlib/camlinternalOO.ml [hunk 1]
```diff
@@ -271,7 +276,7 @@ let new_variable table name =
     index
 
 let to_array arr =
-  if arr = Obj.magic 0 then [||] else arr
+  if arr = Sys.opaque_identity (Obj.magic 0) then [||] else arr
```
Comment (xclerc) [inline ocaml/stdlib/camlinternalOO.ml:279]: I am certainly out of my depth here,
but was wondering why `=` was used
(as opposed to `==`)?

Comment (mshinwell) [reply ocaml/stdlib/camlinternalOO.ml:279]: I don't know; I just followed the existing code!

Comment (xclerc) [reply ocaml/stdlib/camlinternalOO.ml:279]: Oh, indeed, it was not intended as a comment
about the PR, just genuine curiosity.

File: ocaml/stdlib/camlinternalOO.ml [hunk 2]
```diff
@@ -16,13 +16,18 @@
...
-[@@@ocaml.flambda_o3]
-
 [@@@ocaml.inline 0]
 [@@@ocaml.afl_inst_ratio 0]
 
 open Obj
 
+(* Prevent confusion between arrays and blocks *)
+module Array = struct
+  include Array
```
Comment (stedolan) [inline ocaml/stdlib/camlinternalOO.ml:26]: The 'include Array' seems dubious here. I think what you're aiming for is to provide functions called 'Array.unsafe_set' and 'Array.unsafe_get' that can safely be used on blocks, but the 'include Array' also brings in 'Array.get' and 'Array.set' (which, if I understand correctly, are unsafe in this context).

Rather than opaque_identity, should these functions be Obj.field? What type are they used at?

Comment (mshinwell) [reply ocaml/stdlib/camlinternalOO.ml]: These appear to be used at type `array`, but I'm not sure if the underlying value is actually an array -- there is so much `Obj.magic` around that it's hard to follow.

**Comment by mshinwell** [comment]
@stedolan I've pushed a revised version.  This also removes `open Obj` which should make things clearer.

File: ocaml/stdlib/camlinternalOO.ml [hunk 3]
```diff
@@ -16,12 +16,22 @@
...
 open! Stdlib
 
-[@@@ocaml.flambda_o3]
-
 [@@@ocaml.inline 0]
 [@@@ocaml.afl_inst_ratio 0]
 
-open Obj
+let magic x = Sys.opaque_identity (Obj.magic x)
+let obj x = Sys.opaque_identity (Obj.obj x)
```
Comment (stedolan) [inline ocaml/stdlib/camlinternalOO.ml:23]: Maybe call this of_repr or unrepr or something? `obj` is an extremely common variable name in the code below and it's confusing.

Comment (mshinwell) [reply ocaml/stdlib/camlinternalOO.ml]: Fixed

**Review by stedolan — APPROVED** [review]
